spec/services/recaptcha_service_spec.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+describe RecaptchaService do
+  let(:failed_attempts) { 0 }
+  let(:user) { create(:user, failed_attempts: failed_attempts) }
+
+  describe '#recaptcha_configured?' do
+    subject { described_class.recaptcha_configured? }
+
+    let(:site_key) { '11111' }
+    let(:secret_key) { '22222' }
+    let(:configuration) {
+      instance_double(
+        Recaptcha::Configuration,
+        site_key: site_key,
+        secret_key: secret_key,
+      )
+    }
+
+    before do
+      allow(Recaptcha).to receive(:configuration).and_return(configuration)
+    end
+
+    context 'when site key and secret key are present' do
+      it { is_expected.to be true }
+    end
+
+    context 'when site key is blank' do
+      let(:site_key) {}
+
+      it { is_expected.to be false }
+    end
+
+    context 'when secret key is blank' do
+      let(:secret_key) {}
+      
+      it { is_expected.to be false }
+    end
+  end
+
+  describe '#recaptcha_required_for_login?' do
+    subject { described_class.new(user).recaptcha_required_for_login? }
+
+    context 'recaptcha is not configured' do
+      before do
+        allow(described_class).to receive(:recaptcha_configured?).and_return(false)
+      end
+
+      it { is_expected.to be false }
+    end
+
+    context 'when recaptcha is configured' do
+      before do
+        allow(described_class).to receive(:recaptcha_configured?).and_return(true)
+      end
+
+      context 'when the user has no failed logins' do
+        it { is_expected.to be false }
+      end
+
+      context 'when the user has less than 3 failed logins' do
+        let(:failed_attempts) { 2 }
+
+        it { is_expected.to be false }
+      end
+
+      context 'when the user has 3 failed logins' do
+        let(:failed_attempts) { 3 }
+
+        it { is_expected.to be true }
+      end
+
+      context 'when the user has more than 3 failed logins' do
+        let(:failed_attempts) { 4 }
+
+        it { is_expected.to be true }
+      end
+    end
+  end
```

Comment:
You can ignore this as this is just a personal opinion of mine: I suppose this does follow [Better Spec convention](https://www.betterspecs.org/#contexts) although there is a [topic of debate](https://stackoverflow.com/a/51182031/9463774) in regards to too many nested context blocks. I personally am not too much of a fan of it for these case and rather just have `it 'passes if user has more than 3 failed logins'` since it seems short enough to me and nesting seem overkill for single cases, but that's just a matter of opinion. In this case, I prefer to disagree with better specs, but again, just an opinion.

Reply:
Hmm.. yeah I have a tendency to go back and forth on nesting contexts, which I've noticed gets me into trouble from a consistency standpoint sometimes, so lately I've been embracing them. In this case, I see your point. Having it a bit more explicit makes it easier to follow what's happening. I also don't know how I feel about using `is_expected` rather than an explicit `expect`. I didn't even know it was a thing until reading through Better Specs. I think I prefer the explicit assertion, but that might just be because `is_expected` is new to me.

Reply:
Oh wait, I just realized that `failed_attempts` is used when creating the user. Are you suggesting I create the user within the spec? Or update the value on the user within the spec? I actually might just leave this one as is.

Reply:
Whatever is the more explicit solution :) . I prefer to be more explicit as well.

Reply:
Yeah, I'll just create a new user in each spec.

Reply:
Ok, I swung the complete opposite way. Should be about as explicit as you can make it now 😂